### PR TITLE
Adapt IndicesClusterStateService to use allocation ids

### PR DESCRIPTION
**Pull request by bleskes**
#12242 introduced a unique id for an assignment of shard to a node. We should use these id's to drive the decisions made by IndicesClusterStateService when processing the new cluster state sent by the master. If the local shard has a different allocation id than the new cluster state, the shard will be removed and a new one will be created. This fixes a couple of subtle bugs,  most notably a node previously got confused if an incoming cluster state had a newly allocated shard in the initializing state and the local copy was started (which can happen if cluster state updates are bulk processed). In that case, the node have previously re-used the local copy instead of initializing a new one.

 Also, as set of utility methods was introduced on ShardRouting to do various types of matching with other shard routings, giving control about what exactly should be matched (same shard id, same allocation id, all but version and shard info etc.). This is useful here, but also prepares the grounds for the change needed in #12387 (making ShardRouting.equals be strict and perform exact equality).

### CODE REVIEW

**Comment by bleskes** [comment]
@kimchy @imotov  can you have a look?

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 1]
```diff
@@ -526,22 +526,20 @@ public boolean apply(@Nullable RecoveryStatus status) {
...
-                        logger.debug("{} reinitialize shard on primary promotion", indexShard.shardId());
-                        indexService.removeShard(shardId, "promoted to primary");
-                    } else {
-                        // if we happen to remove the shardRouting by id above we don't need to jump in here!
-                        indexShard.updateRoutingEntry(shardRouting, event.state().blocks().disableStatePersistence() == false);
-                    }
+
+                if (shardHasBeenRemoved == false) {
+                    // shadow replicas do not support primary promotion. The master would reinitialize the shard, giving it a new allocation, meaning we should be there.
+                    // nocommit: double check check this
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: @dakrone can you please verify the above?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: The previous behavior was that the service removed the shard if it was a shadow replica being promoted to primary, but it looks like this behavior was removed. Is it moved somewhere else?

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 2]
```diff
@@ -420,7 +420,7 @@ void initialize(String nodeId) {
     void relocate(String relocatingNodeId) {
         ensureNotFrozen();
         version++;
-        assert state == ShardRoutingState.STARTED : this;
+        assert state == ShardRoutingState.STARTED : "current shard has to be started inorder to relocated " + this;
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: "current shard has to be started inorder to relocated " -> "current shard has to be started in order to be relocated "

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java (hunk 1 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: Okay, I tested this locally, it looks like it works differently with the line:

```
[2015-07-22 08:54:10,744][DEBUG][org.elasticsearch.indices.cluster] [node_t1] [test][0] removing shard (different instance of it allocated on this node, current [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [R], v[4], s[STARTED], a[id=93HrpsfMRWybeCn-NxhBjw]], global [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [P], v[6], s[INITIALIZING], a[id=g-oZ-7uiRTmDxgLnTdQHjg], unassigned_info[[reason=REINITIALIZED], at[2015-07-22T14:54:10.743Z]]])[2015-07-22 08:54:10,744][DEBUG][org.elasticsearch.indices.cluster] [node_t1] [test][0] removing shard (different instance of it allocated on this node, current [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [R], v[4], s[STARTED], a[id=93HrpsfMRWybeCn-NxhBjw]], global [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [P], v[6], s[INITIALIZING], a[id=g-oZ-7uiRTmDxgLnTdQHjg], unassigned_info[[reason=REINITIALIZED], at[2015-07-22T14:54:10.743Z]]])
```

Is my understanding correct?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: yeah, my reasoning (which I wanted double checked) is that the master always generates a new allocation id in this case and thus the shard is cleaned by the allocation id check. No need to have special shadow replica handling (but I did add an assert and a comment).

File: core/src/test/java/org/elasticsearch/cluster/routing/ShardRoutingTests.java [hunk 3]
```diff
@@ -19,10 +19,12 @@
 
 package org.elasticsearch.cluster.routing;
 
+import com.carrotsearch.randomizedtesting.annotations.Repeat;
```
Comment (imotov) [inline core/src/test/java/org/elasticsearch/cluster/routing/ShardRoutingTests.java]: Debugging leftovers

**Comment by imotov** [comment]
Looks good to me. I have also rebased  #12387 on this change and it no longer fails.

**Comment by bleskes** [comment]
thx @dakrone @imotov . pushed to master...

**Pull request closed without merging**